src/core/server/saved_objects/migrationsv2/index.ts

```diff
@@ -0,0 +1,240 @@
+        stateP.source = source;
+        stateP.target = `.kibana_${stateP.kibanaVersion}_001`;
+      } else {
+        // TODO: Check if we need to migrate from a v6.x `.kibana` index, source = '.kibana'
+        // This cluster doesn't have any existing Kibana indices, create a new
+        // version specific index.
+        stateP.controlState = 'INIT_NEW_INDICES';
+        stateP.target = `.kibana_${stateP.kibanaVersion}_001`;
+      }
+    }
+    return stateP;
+  } else if (control === 'SET_SOURCE_WRITE_BLOCK') {
+    if ('setIndexWriteBlock' in res) {
+      stateP.controlState = 'CLONE_SOURCE';
+    }
+    return stateP;
+  } else if (control === 'CLONE_SOURCE') {
+    if ('cloneIndex' in res) {
+      stateP.controlState = 'DONE';
+    }
+    return stateP;
+  } else if (control === 'DONE' || control === 'FATAL' || control === 'INIT_NEW_INDICES') {
+    return stateP;
+  } else {
+    return throwBadControlState(control);
+  }
+};
+
+export const next: NextAction = (client: ElasticsearchClient, state: State) => {
+  // If for every state there is only one action (deterministic) we can define
+  // next as a state -> action set.
+  const delay = <F extends () => any>(fn: F): (() => ReturnType<F>) => {
+    return () => {
+      return state.retryDelay > 0
+        ? new Promise((resolve) => setTimeout(resolve, state.retryDelay)).then(fn)
+        : fn();
+    };
+  };
+  const control = state.controlState;
+  if (control === 'INIT') {
+    return delay(Actions.fetchAliases(client, ['.kibana', '.kibana_current', '.kibana_7.11.0']));
+  } else if (control === 'SET_SOURCE_WRITE_BLOCK') {
+    return delay(Actions.setIndexWriteBlock(client, state.source));
+  } else if (control === 'CLONE_SOURCE') {
+    return delay(Actions.cloneIndex(client, state.source, state.target));
+  } else if (control === 'FATAL' || control === 'INIT_NEW_INDICES') {
+    return null;
+  } else if (control === 'DONE') {
+    return null; // DONE
+  } else {
+    return throwBadControlState(control);
+  }
+};
+
+/**
+ * A state-action machine for performing Saved Object Migrations.
+ *
+ * Based on https://www.microsoft.com/en-us/research/uploads/prod/2016/12/Computation-and-State-Machines.pdf
+ *
+ * The state-action machine defines it's behaviour in steps. Each step is a
+ * transition from a state s_i to the state s_i+1 caused by an action a_i.
+ *
+ * s_i   -> a_i -> s_i+1
+ * s_i+1 -> a_i+1 -> s_i+2
+ *
+ * Given a state s1, `next(s1)` returns the next action to execute. Actions are
+ * asynchronous, once the action resolves, we can use the action response to
+ * determine the next state to transition to as defined by the function
+ * `model(state, response)`.
+ *
+ * We can then loosely define a step as:
+ * s_i+1 = model(s_i, await next(s_i)())
+ *
+ * When there are no more actions returned by `next` the state-action machine
+ * terminates.
+ */
+export async function migrationStateMachine(client: ElasticsearchClient, kibanaVersion: string) {
```

Comment:
Why don't we use any existing library implementing `fsm` or state pattern?

Reply:
Finite State Machines use only input and the current control state to determine the next control state. So you have to keep any other state outside of the "machine". The Javascript libraries I looked at also didn't implement a pattern for resolving async side-effects. So it seems like you need to implement that yourself, and then you can synchronously call `myFsm.transition(newState)` but then the total behaviour of the system is no longer contained in the machine.

I did find statem https://medium.com/@venkatperi/javascript-state-machines-a-tutorial-972863e37825 which is more powerful but it only has one contributor which is risky for such a large dependency https://github.com/venkatperi/statem/graphs/contributors I like how they separate "state" and "data" and the pattern matching makes the business logic code you write more succinct. Although data is immutable and needs to be changed using https://github.com/kolodny/immutability-helper which adds overhead and I don't see a lot of benefit for this state machine.